jupyter_scheduler/task_runner.py

```diff
@@ -0,0 +1,310 @@
+            should_compute_run_time = True
+
+        if active_changed and model.active:
+            should_compute_run_time = True
+
+        if should_compute_run_time:
+            next_run_time = self.compute_next_run_time(schedule, timezone)
+        else:
+            next_run_time = cache.next_run_time
+
+        if should_compute_run_time:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(
+                    timezone=timezone, next_run_time=next_run_time, active=active, schedule=schedule
+                ),
+            )
+            self.queue.push(
+                JobDefinitionTask(
+                    job_definition_id=model.job_definition_id, next_run_time=next_run_time
+                )
+            )
+        elif active_changed and not active:
+            self.cache.update(model.job_definition_id, UpdateJobDefinitionCache(active=active))
+        elif schedule_changed or timezone_changed:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(timezone=timezone, schedule=schedule),
+            )
+
+    def delete_job_definition(self):
+        pass
+
+    def pause_jobs(self, job_definition_id: str):
+        self.cache.update(job_definition_id, UpdateJobDefinitionCache(active=False))
+
+    def resume_jobs(self, job_definition_id: str):
+        definition = self.cache.get(job_definition_id)
+        next_run_time = self.compute_next_run_time(definition.schedule, definition.timezone)
+        self.cache.update(
+            job_definition_id, UpdateJobDefinitionCache(active=True, next_run_time=next_run_time)
+        )
+        self.queue.push(
+            JobDefinitionTask(
+                job_definition_id=definition.job_definition_id, next_run_time=next_run_time
+            )
+        )
+
+    def create_job(self, job_definition_id: str):
+        definition = self.scheduler.get_job_definition(job_definition_id)
+        if definition and definition.active:
+            self.scheduler.create_job(
+                CreateJob(**definition.dict(exclude={"schedule", "timezone"}, exclude_none=True))
+            )
+
+    def compute_time_diff(self, queue_run_time: int, timezone: str):
+        local_time = get_localized_timestamp(timezone) if timezone else get_utc_timestamp()
+        return local_time - queue_run_time
+
+    def process_queue(self):
+        while not self.queue.isempty():
+            task = self.queue.peek()
+            cache = self.cache.get(task.job_definition_id)
+            cache_run_time = cache.next_run_time
+            queue_run_time = task.next_run_time
+
+            if not cache.active or queue_run_time != cache_run_time:
+                self.queue.pop()
+                continue
+
+            time_diff = self.compute_time_diff(queue_run_time, cache.timezone)
+            print(f"time difference is : {(time_diff)/1000}")
```

Comment:
For clarity, I suggest we make the following changes:

1. When comparing times, I suggest we always convert them to UTC first for clarity. This logic converts the current time to the task's specified time zone, which I think is much more confusing (although equivalent). This means we can delete `get_localized_timestamp()` and instead just convert `task.next_run_time` to UTC.
2. It's very confusing that the default timezone (UTC) is set by `TaskRunner.compute_time_diff()` in a ternary statement. When setting default values for records, we should set them as "high up" as possible for visibility. I suggest making job definition timezones non-optional and enforcing the default value in either the `JobDefinition` table model or in the `CreateJobDefinition` request model.

Reply:
For 1), I am not sure what you mean, and if that will work with the lib `croniter`, we are using here. I spent quite some time making sure this method works with different cron expressions and for timezones; we should add some tests to make sure this doesn't regress before making changes here.

2) is reasonable, does UTC have an equivalent named timezone?

Reply:
1. I'm saying I personally find this implementation confusing. For time comparisons, I would have expected the job time to be converted to the server's time zone, or for both of them to be converted to UTC. Not sure how else to rephrase this. Because your implementation is equivalent and my concerns are developer-facing, I won't block it, but I do urge your consideration.

2. `Etc/UTC`.

Reply:
Let's discuss this and iterate in another PR; I have other questions around this.